runtime/src/main/java/org/corfudb/util/Utils.java

```diff
@@ -474,36 +494,45 @@ public static void printLogAnatomy(CorfuRuntime runtime, ILogData logData) {
      * @param layout  Latest layout to get clients to fetch tails.
      * @return The max global log tail obtained from the log unit servers.
      */
-    public static long getMaxGlobalTail(Layout layout, CorfuRuntime runtime) {
-        long maxTokenRequested = Address.NON_ADDRESS;
+    public static Tails getTails(Layout layout, CorfuRuntime runtime) {
+        Set<TailsResponse> luResponses = new HashSet<>();
+
         Layout.LayoutSegment segment = layout.getLatestSegment();
 
         // Query the tail of the head log unit in every stripe.
         if (segment.getReplicationMode().equals(Layout.ReplicationMode.CHAIN_REPLICATION)) {
             for (Layout.LayoutStripe stripe : segment.getStripes()) {
-                maxTokenRequested = Math.max(maxTokenRequested,
-                        CFUtils.getUninterruptibly(
-                                runtime.getLayoutView().getRuntimeLayout(layout)
-                                        .getLogUnitClient(stripe.getLogServers().get(0))
-                                        .getTail()));
 
+                TailsResponse res = CFUtils.getUninterruptibly(
+                        runtime.getLayoutView().getRuntimeLayout(layout)
+                                .getLogUnitClient(stripe.getLogServers().get(0))
+                                .getTail());
+                luResponses.add(res);
             }
         } else if (segment.getReplicationMode()
                 .equals(Layout.ReplicationMode.QUORUM_REPLICATION)) {
             for (Layout.LayoutStripe stripe : segment.getStripes()) {
-                CompletableFuture<Long>[] completableFutures = stripe.getLogServers()
+                CompletableFuture<TailsResponse>[] completableFutures = stripe.getLogServers()
                         .stream()
                         .map(s -> runtime.getLayoutView().getRuntimeLayout(layout)
                                 .getLogUnitClient(s).getTail())
                         .toArray(CompletableFuture[]::new);
-                QuorumFuturesFactory.CompositeFuture<Long> quorumFuture =
-                        QuorumFuturesFactory.getQuorumFuture(Comparator.naturalOrder(),
-                                completableFutures);
-                maxTokenRequested = Math.max(maxTokenRequested,
-                        CFUtils.getUninterruptibly(quorumFuture));
 
+                QuorumFuturesFactory.CompositeFuture<TailsResponse> quorumFuture =
+                        QuorumFuturesFactory.getQuorumFuture(new TailsResponseComparator(),
+                                completableFutures);
+                TailsResponse res = CFUtils.getUninterruptibly(quorumFuture);
+                luResponses.add(res);
             }
         }
-        return maxTokenRequested;
+        return getTails(luResponses);
+    }
+
+    static class TailsResponseComparator implements Comparator<TailsResponse> {
```

Comment:
Please provide more comments on what this comparator is used for. As far as I can tell, as long as you have a majority you should be able to get max values as all majorities intersect.

Reply:
Removed it.